staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go

```diff
@@ -146,14 +146,29 @@ type Store struct {
 	// specific cases where storage of the value is not appropriate, since
 	// they cannot be watched.
 	Decorator ObjectFunc
+
 	// CreateStrategy implements resource-specific behavior during creation.
 	CreateStrategy rest.RESTCreateStrategy
+	// BeginCreate is an optional hook that returns a "transaction-like"
+	// commit/cleanup function which will be called at the end of operation,
+	// whether it succeeds or not.  If the overall operation panics, the
+	// cleanup function will not be called.  Almost nobody should use this
```

Comment:
We could change this, probably - instead of passing err to the cleanup, we could simply pass a bool.  Declare that bool JUST for this purpose and defer a call to the cleanup.

Altertnately we could define a Transaction interface with "Commit" and "Revert" methods, where we define Revert() after Commit() to be a no-op, then defer the Revert, and call Commit() in the success path.  More involved but less tricky?

Happy to explore either to alleviate this caveat.

Reply:
I like constraining the callback. Commit()/Close() is appealing at first blush